pimd/pim6_mld.c

```diff
@@ -442,7 +442,7 @@ static void gm_sg_update(struct gm_sg *sg, bool has_expired)
 
 	} else if (sg->tib_joined && !new_join) {
 		tib_sg_gm_prune(gm_ifp->pim, sg->sgaddr, gm_ifp->ifp, &sg->oil);
-
+		pim_channel_oil_del(sg->oil, __func__);
```

Comment:
@eqvinox when prune is received for IPV4, channel_oil is freed twice since the reference counter is 2.

**1st case:**
igmp_group_timer
igmp_anysource_forward_stop
igmp_source_forward_stop
tib_sg_gm_prune
pim_ifchannel_local_membership_del
delete_on_noinfo
pim_ifchannel_delete
pim_upstream_del
upstream_channel_oil_detach
pim_channel_oil_upstream_deref
pim_channel_oil_del

**2nd case:**
igmp_group_timer
igmp_source_delete_expired
igmp_source_delete
source_channel_oil_detach
pim_channel_oil_del

**For IPV6**
1st will be deleted in tib_sg_gm_prune which will deref the channel_oil from upstream.
2nd, we need to deref the channel_oil from sg.

Reply:
When we receive MLD join, we are doing the below stuff.
static struct channel_oil *
tib_sg_oil_setup(struct pim_instance *pim, pim_sgaddr sg, struct interface *oif)
{
        struct pim_interface *pim_oif = oif->info;
        int input_iface_vif_index = 0;
        pim_addr vif_source;
        struct prefix src, grp;
        struct pim_nexthop nexthop;
        struct pim_upstream *up = NULL;

        if (!pim_rp_set_upstream_addr(pim, &vif_source, sg.src, sg.grp)) {
                /* no PIM RP - create a dummy channel oil */
                return pim_channel_oil_add(pim, &sg, __func__);
        }